HyperVExtension/src/HyperVExtension/Services/VMGalleryService.cs

```diff
@@ -0,0 +1,129 @@
+using Windows.Storage;
+
+namespace HyperVExtension.Services;
+
+/// <summary>
+/// Service that performs operations specific to the VM gallery.
+/// </summary>
+public sealed class VMGalleryService : IVMGalleryService
+{
+    private const string ComponentName = "VMGalleryService";
+
+    private readonly IDownloaderService _downloaderService;
+
+    private readonly JsonSerializerOptions _jsonOptions = new()
+    {
+        PropertyNameCaseInsensitive = true,
+        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+        TypeInfoResolver = JsonSourceGenerationContext.Default,
+        AllowTrailingCommas = true,
+    };
+
+    public VMGalleryService(IDownloaderService downloaderService)
+    {
+        _downloaderService = downloaderService;
+    }
+
+    private static readonly Uri VmGalleryUrl = new("https://go.microsoft.com/fwlink/?linkid=851584");
+
+    private VMGalleryImageList _imageList = new();
+
+    /// <inheritdoc cref="IVMGalleryService.GetGalleryImagesAsync"/>
+    public async Task<VMGalleryImageList> GetGalleryImagesAsync()
+    {
+        // If we have already downloaded the image list before, return it.
+        if (_imageList.Images.Count > 0)
+        {
+            return _imageList;
+        }
+
+        var emptyList = new VMGalleryImageList();
+        try
+        {
+            var cancellationTokenSource = new CancellationTokenSource();
+
+            // This should be quick as the file is around 28.0 KB, so a 5 minute timeout should be ok on the worst network.
+            cancellationTokenSource.CancelAfter(TimeSpan.FromMinutes(5));
+
+            // get the JSON data
+            var resultJson = await _downloaderService.DownloadStringAsync(VmGalleryUrl.AbsoluteUri, cancellationTokenSource.Token);
+            _imageList = JsonSerializer.Deserialize(resultJson, typeof(VMGalleryImageList), _jsonOptions) as VMGalleryImageList ?? emptyList;
+
+            // Now we need to download the base64 images for the symbols (icons). So they can be used within an adaptive card.
+            foreach (var image in _imageList.Images)
+            {
+                if (!string.IsNullOrEmpty(image.Symbol.Uri))
+                {
+                    var byteArray = await _downloaderService.DownloadByteArrayAsync(image.Symbol.Uri, cancellationTokenSource.Token);
+                    if (!ValidateFileSha256Hash(byteArray, image.Symbol.Hash))
+                    {
+                        Logging.Logger()?.ReportError(ComponentName, $"Symbol Hash '{image.Symbol.Hash}' validation failed for image with name '{image}'. Symbol uri: '{image.Symbol}'");
+                        continue;
+                    }
+
+                    image.Symbol.Base64Image = Convert.ToBase64String(byteArray);
+                }
+            }
+        }
+        catch (Exception ex)
+        {
+            Logging.Logger()?.ReportError(ComponentName, $"Unable to retrieve VM gallery images", ex);
+        }
+
+        return _imageList;
```

Comment:
will `_imageList` ever be empty at this point?  What happens if an exception happens on line `56`?

Reply:
_imageList being empty at this point is fine. The caller can check if the list is empty. (We as in Microsoft and likely our team soon) own this Json file, It shouldn't ever be empty but I handle the case if it is, in the caller.